### PR TITLE
Generalized lemma broadcast, implements #37

### PR DESCRIPTION
**Pull request by utaal-b**
This implements #37:
* a mechanism to produce quantified axioms from regular lemmas, `broadcast proof fn`s,
* an accompanying mechanism to import these quantified facts (`broadcast use`) at either the module or function level, and
* a mechanism to construct groups of `broadcast proof fn`s that can be `broadcast use`d (`broadcast group`).

The underlying implementation uses the `fuel` mechanism that was already implemented for `broadcast_forall`, but extends it to allow the construction of groups and the ability to `broadcast use` at the module level.

Here's an example:
```rust
    mod ring {
        use builtin::*;

        pub struct Ring {
            pub i: nat,
        }

        impl Ring {
            pub closed spec fn inv(&self) -> bool {
                self.i < 10
            }

            pub closed spec fn succ(&self) -> Ring {
                Ring { i: if self.i == 9 { 0 } else { self.i + 1 } }
            }

            pub closed spec fn prev(&self) -> Ring {
                Ring { i: if self.i == 0 { 9 } else { (self.i - 1) as nat } }
            }

            pub broadcast proof fn succ_ensures(p: Ring)
                requires p.inv()
                ensures p.inv() && (#[trigger] p.succ()).prev() == p
            { }

            pub broadcast proof fn prev_ensures(p: Ring)
                requires p.inv()
                ensures p.inv() && (#[trigger] p.prev()).succ() == p
            { }

            pub broadcast group properties {
                Ring::succ_ensures,
                Ring::prev_ensures,
            }
        }
    }

    mod m2 {
            use builtin::*;
            use crate::ring::*;

            proof fn t2(p: Ring) requires p.inv() {
                broadcast use Ring::succ_ensures;
                assert(p.succ().prev() == p);
            }

            proof fn t3(p: Ring) requires p.inv() {
                broadcast use Ring::succ_ensures;
                assert(p.succ().prev() == p);
                assert(p.prev().succ() == p); // FAILS
            }

            proof fn t4(p: Ring) requires p.inv() {
                assert(p.prev().succ() == p); // FAILS
            }

            proof fn t5(p: Ring) requires p.inv() {
                broadcast use Ring::succ_ensures, Ring::prev_ensures;
                assert(p.succ().prev() == p);
                assert(p.prev().succ() == p);
            }

            proof fn t6(p: Ring) requires p.inv() {
                broadcast use Ring::properties;
                assert(p.succ().prev() == p);
                assert(p.prev().succ() == p);
            }
    }
```

A significant part of this PR are syntax additions.

The most important from a soundness perspective are the additional function graph nodes and edges that represent groups and module-level reveals, to prevent circular reasoning. In particular:

https://github.com/verus-lang/verus/pull/1022/files#diff-b7fce60d4dba89afe470adb93f36bbc6d33da444e711c78c476970e52dee9105R322-R351

https://github.com/verus-lang/verus/blob/852db350210aef39cadbb944f6445396d39795ce/source/vir/src/context.rs#L322-L351

and

https://github.com/verus-lang/verus/pull/1022/files#diff-b7fce60d4dba89afe470adb93f36bbc6d33da444e711c78c476970e52dee9105R602-R622

https://github.com/verus-lang/verus/blob/852db350210aef39cadbb944f6445396d39795ce/source/vir/src/context.rs#L602-L622

The PR also introduces two attributes for broadcast groups:

* `#[verifier::hidden_unless_this_module_is_used]` which instructs prune to remove the broadcasted axioms if this broadcast group was `use`d indirectly but nothing else in the module was used (this is to preserve the existing pruning mechanism in `vstd`);

* `#[verifier::broadcast_use_by_default_when_this_crate_is_imported]` which can be used to tag a default broadcast group for a crate (currently intended for vstd) that will be automatically `broadcast use`d in dependent crates.

---

A follow-on PR, https://github.com/verus-lang/verus/pull/1023, uses this mechanism to replace the existing, unconditional broadcast via `#[verifier::broadcast_forall] #[verifier::external_body]`.

### CODE REVIEW

**Review by tjhance — COMMENTED** [review]
> #[verifier::hidden_unless_this_module_is_used] which instructs prune to remove the broadcasted axioms if this broadcast group was used indirectly but nothing else in the module was used (this is to preserve the existing pruning mechanism in vstd);

I don't get what this means. What does it mean for a group to be used "indirectly"? What's an example where this changes behavior?

**Comment by utaal-b** [comment]
> What does it mean for a group to be used "indirectly"? What's an example where this changes behavior?

That could have been phrased better (by me). A crate has a `#[verifier::broadcast_use_by_default_when_this_crate_is_imported]` which will `broadcast use` a default group. For vstd, this will be the equivalent group to the current `broadcast_forall` axioms, but those get pruned if nothing else in the module they appear in was used, so `#[verifier::hidden_unless_this_module_is_used]` will instruct the pruner to prune the `broadcast proofs` in this group unless the current module is used (to preserve the current behavior for the `vstd` axioms).

File: source/rust_verify/src/rust_to_vir.rs [hunk 1]
```diff
@@ -82,6 +83,89 @@ fn check_item<'tcx>(
...
         }
+        if vattrs.item_broadcast_use {
+            let err = crate::util::err_span(item.span, "invalid module-level broadcast use");
+            let ItemKind::Const(_ty, generics, body_id) = item.kind else {
+                return err;
+            };
+            unsupported_err_unless!(
+                generics.params.len() == 0 && generics.predicates.len() == 0,
+                item.span,
+                "const generics"
```
[the remark this answers is not in the document]
Comment (utaal-b) [reply source/rust_verify/src/rust_to_vir.rs]: Done.

**Review by tjhance — APPROVED** [review]

**Comment by utaal-b** [comment]
This will be mergeable once https://github.com/verus-lang/verusfmt/pull/48 is merged so that `verusfmt` passes.

**Comment by jaybosamiya** [comment]
https://github.com/verus-lang/verusfmt/pull/48 is merged (thanks for making the PR @utaal-b!), and new version of verusfmt (v0.2.10) is released, so the CI should now pass here. Triggered a re-run of the failed CI, hopefully should be a green!

**Pull request merged**